packages/terra-navigation-prompt/docs/README.md

```diff
@@ -0,0 +1,138 @@
+
+Another solution is to expose callbacks from the stateful components. Whenever the stateful component's state changes, the callbacks are executed to notify the consumer of the current state. This works well at a small scale, but this becomes hard to maintain as the component tree increases in depth. One would have to chain these callbacks many levels deep to ensure every component is kept abreast of the current state, and every component, even those without these state concerns, would have to adhere to a consistent prop interface for those callbacks.
+
+The NavigationPrompt and the NavigationPromptCheckpoint were created smooth over the rough edges of the aforementioned strategies. Components that have sensitive state can render a NavigationPrompt to make their state known to the application. Components that navigate between components can use the NavigationPromptCheckpoint to check for the presence of NavigationPrompts and inform the user in a consistent way. State remains localized to the component that needs it, communication of state changes occurs automatically, and components that don't share these concerns can safely ignore all of it.
+
+## Usage
+
+### `NavigationPrompt`
+
+When a component renders a NavigationPrompt, it is telling its parent components two things: that the component is in the middle of something important, and that users may care about the component going away.
+
+The NavigationPrompt can be rendered with two props: a `description` string that gives context to the NavigationPrompt, and a `metaData` object that can be used to provide additional data that may be relevant to the NavigationPrompt.
+
+The NavigationPrompt can be rendered like any other component, but prop consistency should be maintained between renders when possible. Creating a new object for the `metaData` prop within a render function will cause the NavigationPrompt to deregister and reregister itself during every update:
+
+```jsx
+render() {
+  return (
+    <NavigationPrompt description="This prompt will register and notify every time the component renders" metaData={{ important: 'info' }} />
+  );
+}
+```
+
+To avoid this, the `metaData` value can be stored inside a component's state or an internal property and updated only when necessary:
+
+```jsx
+constructor() {
+  this.state = {
+    metaDataValue: {
+      important: 'info',
+    }
+  }
+}
+
+render() {
+  return (
+    <NavigationPrompt description="My props are consistent between renders, so this prompt will not re-register or notify" metaData={this.state.metaDataValue} />
+  );
+}
+```
+
+### `NavigationPromptCheckpoint`
+
+The NavigationPromptCheckpoint can be used to detect the presence of NavigationPrompts rendered within it. 
```

Comment:
playing dumb, this should be the parent component that should implement this? Or top level navigation or application root?

Reply:
Any component that navigates between stateful child components should render NavigationPromptCheckpoints around those children. Each NavigationPrompt rendered by a child will be registered to the NavigationPromptCheckpoints above it in the component tree. The implementer of each NavigationPromptCheckpoint can use the registered data to prompt the user before navigating or prevent navigation altogether.